File: front-constructora/src/utils/validations.ts
// Validacion para nombres y apellidos
export const validateName = (name: string) => {
  const nameRegex = /^[a-zA-ZáéíóúÁÉÍÓÚñÑ\s']+$/;

  if (!name.trim()) return "Este campo es requerido";
  if (!nameRegex.test(name)) return "Solo se permiten letras y espacios";
  if (name.trim().length < 2) return "Debe tener al menos 2 caracteres";
  if (name.trim().length > 50) return "No puede exceder los 50 caracteres";
  return "";
};

//Validacion para username
export const validateUsername = (username: string) => {
  const usernameRegex = /^[a-zA-ZñÑ][a-zA-Z0-9ñÑ]*(?:[-_][a-zA-ZñÑ0-9]+)*$/;

  if (!username) return "Este campo es requerido";
  if (!usernameRegex.test(username)) {
    return (
      "Solo letras (a-z, A-Z), números y guiones (-_)\n" +
      "No puede empezar con números o guiones\n" +
      "No puede terminar con guiones"
    );
  }
  if (username.length < 4) return "Mínimo 4 caracteres";
  if (username.length > 30) return "Maximo 20 caracteres";

  return "";
};

// Validacion para contraseña
export const validatePassword = (password: string) => {
  if (!password) return "Este campo es requerido";
  if (password.length < 4) return "Debe tener al menos 4 caracteres";
  if (password.length > 16) return "No puede exceder los 16 caracteres";
  return "";
};

// Validacion para telefono en (México)
export const validatePhoneMX = (phone: string) => {
  if (!phone.trim()) return "Este campo es requerido";
  if (/\D/.test(phone)) return "Solo se permiten números (0-9)";
  if (phone[0] === "0" || phone[0] === "1") return "La region no es valida";
  if (phone.length !== 10) return "Debe tener 10 dígitos exactos";
  return "";
};

// Validación para rol
export const validateRole = (role: string) => {
  if (!role) return "Este campo es requerido";
  return "";
};

// Validacion para direccion
export const validateAddress = (address: string) => {
  if (!address.trim()) return "Este campo es requerido";
  return "";
};

// Validación para fk(number != 0)
export const validateFk = (fk: number) => {
  if (!fk) return "Este campo es requerido";
  return "";
};

export const validateStartDate = (date: string) => {
  if (!date) return "Este campo es requerido";
  return "";
};

export const validateEndDate = (startDate: string, endDate: string) => {
  if (!startDate) return "La fecha de inicio no sea definido";
  if (!endDate) return "Este campo es requerido";

  const start = new Date(startDate);
  const end = new Date(endDate);

  if (end < start) {
    return "La fecha de finalización no puede ser anterior a la de inicio.";
  }

  const diffInMs = end.getTime() - start.getTime();
  const diffInDays = diffInMs / (1000 * 60 * 60 * 24);

  if (diffInDays < 7) {
    return "El proyecto debe tener una duración mínima de una semana.";
  }

  return "";
};

File: back-constructora/src/main/java/com/back_constructora/util/Validations.java
package com.back_constructora.util;
import java.util.regex.Pattern;

import com.back_constructora.model.Role;

public class Validations 
{
    // Validacion para nombres y apellidos
    public static void validateName(String name) 
    {
        String nameRegex = "^[a-zA-ZáéíóúÁÉÍÓÚñÑ\\s']+$";
        
        if (name == null || name.trim().isEmpty()) {
            throw new IllegalArgumentException("$Nombre: es un campo requerido");
        }
        if (!Pattern.matches(nameRegex, name)) {
            throw new IllegalArgumentException("$Nombre: Solo se permiten letras y espacios");
        }
        if (name.trim().length() < 2) {
            throw new IllegalArgumentException("$Nombre: Debe tener al menos 2 caracteres");
        }
        if (name.trim().length() > 50) {
            throw new IllegalArgumentException("$Nombre: No puede exceder los 50 caracteres");
        }
    }

    // Validación para username
    public static void validateUsername(String username) {
        String usernameRegex = "^[a-zA-ZñÑ][a-zA-Z0-9ñÑ]*(?:[-_][a-zA-ZñÑ0-9]+)*$";
        
        if (username == null || username.trim().isEmpty()) {
            throw new IllegalArgumentException("$Nombre de usuario: es un campo requerido");
        }
        System.out.println(username);
        if (!Pattern.matches(usernameRegex, username)) {
            throw new IllegalArgumentException(
                "$Nombre de usuario: Solo letras (a-z, A-Z), números y guiones (-_)\n" +
                "No puede empezar con números o guiones\n" +
                "No puede terminar con guiones");
        }
        if (username.length() < 4) {
            throw new IllegalArgumentException("$Nombre de usuario: Mínimo 4 caracteres");
        }
        if (username.length() > 30) {
            throw new IllegalArgumentException("$Nombre de usuario: Máximo 30 caracteres");
        }
    }

    // Validación para contraseña
    public static void validatePassword(String password) {
        if (password == null || password.isEmpty()) {
            throw new IllegalArgumentException("$Contraseña: es un campo requerid");
        }
        if (password.length() < 4) {
            throw new IllegalArgumentException("$Contraseña: Debe tener al menos 4 caracteres");
        }
        if (password.length() > 16) {
            throw new IllegalArgumentException("$Contraseña: No puede exceder los 16 caracteres");
        }
    }

    // Validación para teléfono en México
    public static void validatePhoneMX(String phone) {
        if (phone == null || phone.trim().isEmpty()) {
            throw new IllegalArgumentException("$Teléfono: Este campo es requerido");
        }
        if (phone.matches(".*\\D.*")) {
            throw new IllegalArgumentException("$Teléfono: Solo se permiten números (0-9)");
        }
        if (phone.startsWith("0") || phone.startsWith("1")) {
            throw new IllegalArgumentException("$Teléfono: La región no es válida");
        }
        if (phone.length() != 10) {
            throw new IllegalArgumentException("$Teléfono: Debe tener 10 dígitos exactos");
        }
    }

    // Validación para rol
    public static void validateRole(Role role) 
    {
        if (role == null) 
        {
            throw new IllegalArgumentException("$Rol: Este campo es requerido");
        }
    }
}
